File: solana/bridge-program/tests/bridge-program.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { BridgeProgram } from "../target/types/bridge_program";

describe("bridge-program", () => {
  // Configure the client to use the local cluster.
  anchor.setProvider(anchor.AnchorProvider.env());

  const program = anchor.workspace.BridgeProgram as Program<BridgeProgram>;

  try {
    it("Is initialized!", async () => {
      // Add your test here.
      //const tx = await program.methods.initialize().rpc();
      //console.log("Your transaction signature", tx);

      const [bridgeAccount] = await anchor.web3.PublicKey.findProgramAddressSync(
        [Buffer.from('bridge_vault')],
        program.programId
      );

      await program.methods.initialize()
      .accounts({ bridgeAccount })
      .rpc()
      console.log("Bridge PDA:", bridgeAccount.toBase58());
    });
  } catch (error) {
    console.error("❌ Initialize failed:", error);
  }

  it("Lock Sol", async () => {
    await program.methods.lockSol(
      new anchor.BN(1_000_000_00),
    )
    .accounts({
      bridgeAccount :bridgePDa,
      user : provider.wallet.PublicKey,
    })
    .rpc();
    
  });
  
});
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { BridgeProgram } from "../target/types/bridge_program";
import { assert } from "chai";

describe("bridge-program", () => {
  // 1. Setup Provider
  const provider = anchor.AnchorProvider.local(); // Better for testing than .env()
  anchor.setProvider(provider);

  // 2. Load Program
  const program = anchor.workspace.BridgeProgram as Program<BridgeProgram>;

  // 3. Declare variables we'll reuse
  let bridgePda;
  let user = provider.wallet.publicKey;

  // 4. Initialize Test
  it("Initialize Bridge Account", async () => {
    // Calculate PDA (matches Rust program)
    [bridgePda] = anchor.web3.PublicKey.findProgramAddressSync(
      [Buffer.from('bridge_vault')],
      program.programId
    );

    // Call initialize instruction
    const tx = await program.methods.initialize()
      .accounts({ bridgeAccount: bridgePda })
      .rpc();

    console.log("✅ Initialized. PDA:", bridgePda.toBase58(), "TX:", tx);

    // Verify initialization
    const account = await program.account.bridgeAccount.fetch(bridgePda);
    assert.equal(account.totalLocked.toNumber(), 0, "Should start with 0 locked");
  });

  // 5. Lock SOL Test
  it("Lock SOL into Bridge", async () => {
    const amount = new anchor.BN(1_000_000_000); // 1 SOL

    const tx = await program.methods.lockSol(amount)
      .accounts({
        bridgeAccount: bridgePda, // Reuse the PDA
        user: user               // Test wallet
      })
      .rpc();

    console.log("✅ Locked", amount.toNumber()/1e9, "SOL. TX:", tx);

    // Verify lock
    const account = await program.account.bridgeAccount.fetch(bridgePda);
    assert.equal(
      account.totalLocked.toNumber(),
      amount.toNumber(),
      "Locked amount should match"
    );
  });
});

File: solana/bridge-program/tests/bridge-program.rs
use anchor_lang::{prelude::*};
use anchor_lang::solana_program::system_program;
use solana_sdk::{signature::Keypair, transaction::Transaction};
use anchor_client::solana_sdk::pubkey::Pubkey;

#[tokio::test]
async fn test_initialize_and_lock_sol() {
    let program_id = Pubkey::from_str("28AQpwDXyQPTkcuJweUQFfAMqTkDZfNME71Anic7o5rM").unwrap();
    let client = anchor_client::Client::new_with_options(
        solana_client::rpc_client::RpcClient::new("https://api.devnet.solana.com".to_string()),
        Keypair::new(),
        anchor_client::ClientOptions::default(),
    );

    let program = client.program(program_id);

    // Step 1: Initialize the bridge account
    let user = Keypair::new();
    let bridge_account = Keypair::new();

    let ix_initialize = program
        .request()
        .accounts(bridge_program::accounts::Initialize {
            bridge_account: bridge_account.pubkey(),
            user: user.pubkey(),
            system_program: system_program::id(),
        })
        .signers(&[&user, &bridge_account])
        .instruction();
    
    let mut tx = Transaction::new_with_payer(&[ix_initialize], Some(&user.pubkey()));
    client.rpc().send_and_confirm_transaction(&tx).unwrap();

    // Verify the bridge account is initialized (optional: check state on-chain)

    // Step 2: Lock some SOL
    let amount = 1000000000; // 1 SOL in lamports
    let ix_lock_sol = program
        .request()
        .accounts(bridge_program::accounts::LockSol {
            bridge_account: bridge_account.pubkey(),
            user: user.pubkey(),
            system_program: system_program::id(),
        })
        .args(bridge_program::instruction::LockSol { amount })
        .signers(&[&user])
        .instruction();
    
    let mut tx = Transaction::new_with_payer(&[ix_lock_sol], Some(&user.pubkey()));
    client.rpc().send_and_confirm_transaction(&tx).unwrap();

    // Optional: Assert that the total_locked in the bridge account has been updated
}
